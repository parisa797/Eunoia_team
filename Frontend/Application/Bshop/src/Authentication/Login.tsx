import React, { useRef , useState } from "react";
// import { TextInput as RNTextInput } from "react-native";
import { CommonActions } from "@react-navigation/native";
import { BorderlessButton } from "react-native-gesture-handler";
import { useFormik } from "formik";
import * as Yup from "yup";
import { Container, Button, Box } from "../components";
import { AuthNavigationProps } from "../components/Navigation";
// import TextInput from "../components/Form/TextInput";
import Checkbox from "../components/Form/Checkbox";
import Footer from "./components/Footer";


import { StatusBar } from "expo-status-bar";
import * as SecureStore from "expo-secure-store";
import FlashMessage from "react-native-flash-message";
import { showMessage, hideMessage } from "react-native-flash-message";
import {
  StyleSheet,
  Text,
  View,
  Image,
  TextInput,
  TouchableOpacity,
  Alert,
  TouchableWithoutFeedback,
  Keyboard,
} from "react-native";
// import { useSelector, useDispatch } from "react-redux";
// const LoginSchema = Yup.object().shape({
//   email: Yup.string().email("Invalid email").required("Required"),
//   password: Yup.string()
//     .min(2, "بسیار کوتاهه! دوباره تلاش کن")
//     .max(50, "خیلی طولانیه! دوباره تلاش کن")
//     .required("Required"),
// });
async function save(key, value) {
  await SecureStore.setItemAsync(key, value);
}


const Login = ({ navigation }: AuthNavigationProps<"Login">) => {
   const [email, setEmail] = useState("");
  const [pass, setPassword] = useState("");

  // const count = useSelector((state) => state.counter.value);
  // const count = useSelector((state) => state.isLoggedin.value);
  // console.log(count);

  // const dispatch = useDispatch();


   const HandleLogin = (email, pass) => {
  // const dispatch = useDispatch();
  console.log("in handle login");
  console.log(email);
  // console.log(dispatch(increment()));
  // if (email == undefined && pass == undefined) {
  if (email.length == 0 && pass.length == 0) {
    showMessage({
      message: "لطفا اطلاعات کاربری خود را وارد کنید.",
      type: "warning",
      backgroundColor: "#f1f1f2", // background color
      color: "#000", // text color
      statusBarHeight: "8",
      titleStyle: {
        fontSize: 15,
      },
    });
  } else {
    if (email.length == 0) {
      // alert("Please enter your email");
      showMessage({
        message: "لطفا ایمیل خود را وارد کنید.",
        type: "warning",
        backgroundColor: "#f1f1f2", // background color
        color: "#000", // text color
        statusBarHeight: "8",
        titleStyle: {
          fontSize: 15,
        },
      });
      // jest.spyOn(Alert, "alert");
    } else {
      if (pass.length == 0) {
        // alert("Please enter your password");
        showMessage({
          message: "لطفا پسورد خود را وارد کنید.",
          type: "warning",
          backgroundColor: "#f1f1f2", // background color
          color: "#000", // text color
          statusBarHeight: "8",
          titleStyle: {
            fontSize: 15,
          },
        });
      } else {
        LoginFetch(email, pass);
      }
    }
  }
}; 
 const LoginFetch = (email, pass) => {
  const requestOptions = {
    method: "POST",
    headers: {
      Accept: "application/json",
      "Content-Type": "application/json",
      "Access-Control-Allow-Origin": "*",
      "Access-Control-Allow-Headers": "*",
    },
    body: JSON.stringify({
      email: email, //change this!!!
      password: pass,
    }),
  };
  fetch("http://10.0.2.2:8000/rest-auth/login/", requestOptions)
    .then((response) => {
      console.log(response);
      console.log(response.status);
      if (response.status == 200) {
        alert("You've logged in successfully");
        
        
      } else alert("Incorrect email or password");

      return response.json();
    })
    .then(async (result) => {
      console.log(result);
      save("token", result.key);
      navigation.dispatch(
        CommonActions.reset({
          index: 0,
          routes: [{ name: "Home" }],
        })
      );
      // for (let key in result) {
      //   switch (key) {
      //     case "non_field_errors": //put different errors
      //       break;

      //     case "key":
      //       save("token", result.key);
      //       break;

      //     default:
      //       console.log(result);
      //   }
      // }
      // save("token", result.key);
    })
    .catch((error) => console.log("error", error));
};
  const pressSignup = () => {
    navigation.navigate("SignUp"); //this line
    //navigation.push("signup")
  };
  return (
    <TouchableWithoutFeedback
      onPress={() => {
        Keyboard.dismiss();
        // console.log("dismissed keyboard");
      }}
    >
      <View style={styles.container}>
        {/* <Image
          style={styles.image}
          source={require("../assets/Login-amico.png")}
        /> */}
        <StatusBar style="auto" />
        <View style={styles.inputView}>
          <TextInput
            // data-testID="usernameInput"
            style={styles.TextInput}
            placeholder="ایمیل"
            placeholderTextColor="#000"
            onChangeText={(email) => setEmail(email)}
          />
        </View>
        <View style={styles.inputView}>
          <TextInput
            style={styles.TextInput}
            placeholder="رمز عبور"
            placeholderTextColor="#000"
            secureTextEntry={true}
            onChangeText={(pass) => setPassword(pass)}
          />
        </View>
        {/* <TouchableOpacity>
        <Text style={styles.forgot_button}>Forgot Password?</Text>
      </TouchableOpacity> */}
        <TouchableOpacity
          style={styles.loginBtn}
          // onPress={() => HandleLogin(email, pass)}
          // onPress={() => dispatch(increment())}
          onPress={() => {
            // login();
            HandleLogin(email, pass);
            // const count = useSelector((state) => state.isLoggedin.value);
            // console.log(count);
          }}
        >
          <Text style={styles.loginText}>ورود</Text>
        </TouchableOpacity>

        <Text style={{ fontSize: 20 }}>حساب کاربری ندارید؟</Text>
        <Text style={{ color: "#b31414", fontSize: 20 }} onPress={pressSignup}>
          ثبت نام
        </Text>

        <FlashMessage position="top" />
      </View>
    </TouchableWithoutFeedback>
  );
};

const styles = StyleSheet.create({
  container: {
    flex: 1,
    backgroundColor: "#fff",
    alignItems: "center",
    justifyContent: "center",
  },

  // image: {
  //   marginBottom: "10%",
  //   // marginTop: "-20%",
  //   width: "80%",
  //   height: "50%",

  //   // position: "absolute",
  //   // marginTop: 10,
  //   // resizeMode: "center",
  //   // marginTop: 20,
  //   // alignContent: "stretch",
  //   // alignSelf: "center",
  // },

  inputView: {
    backgroundColor: "#f1f1f2",
    borderRadius: 20,
    width: "80%",
    height: 45,
    marginBottom: 20,
    alignItems: "center",
    alignContent: "center",
  },

  TextInput: {
    height: 50,
    flex: 1,
    padding: 10,
    marginLeft: 20,
    fontSize: 20,
  },

  // forgot_button: {
  //   height: 30,
  //   marginBottom: 30,
  // },

  loginText: {
    color: "#fff",
    fontSize: 20,
  },

  loginBtn: {
    width: "40%",
    borderRadius: 20,
    height: 50,
    alignItems: "center",
    justifyContent: "center",
    marginTop: 10,
    marginBottom: 20,
    backgroundColor: "#b31414",
  },
});

export default Login;

