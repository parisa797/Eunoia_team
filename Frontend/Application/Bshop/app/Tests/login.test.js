import React from "react";
import renderer from "react-test-renderer";
import { act } from "react-test-renderer";
import Login from "../login";

import {
  render,
  fireEvent,
  cleanup,
  waitFor,
} from "react-native-testing-library";
// import ReactDOM from "react-dom";
// // import { TouchableOpacity } from "react-native";
// import FlashMessage from "react-native-flash-message";
// import { showMessage } from "react-native-flash-message";

// global.fetch = jest.fn();

// // mocking an API success response once
// fetch.mockResponseIsSuccess = (body) => {
// fetch.mockImplementationForOnce(() =>
// Promise.resolve({ json: () => Promise.resolve(JSON.parse(body)) })
// );
// };

// // mocking an API failure response for once
// fetch.mockResponseIsFailure = (error) => {
// fetch.mockImplementationForOnce(() => Promise.reject(error));
// };

describe("Login component be tested", () => {
  // jest.mock(
  //   "react-native/Libraries/Components/Touchable/TouchableOpacity.js",
  //   () => {
  //     const { TouchableHighlight } = require("react-native");
  //     const MockTouchable = (props) => {
  //       return <TouchableHighlight {...props} />;
  //     };
  //     MockTouchable.displayName = "TouchableOpacity";

  //     return MockTouchable;
  //   }
  // );
  // jest.mock(
  //   "react-native/Libraries/Components/Touchable/TouchableOpacity",
  //   () => "TouchableOpacity"
  // );
  //snapshot test
  it("renders correctly", () => {
    const tree = renderer.create(<Login />).toJSON();
    expect(tree).toMatchSnapshot();
  });

  //things to test:
  // email state changes by inputtext
  it("check email input", async () => {
    const push = jest.fn();
    const { getByTestId, getByPlaceholderText } = await render(
      <Login navigation={{ push }} />
    );
    const email = "hi@gmail.com";
    await fireEvent.changeText(getByTestId("email_check"), email);
    expect(getByTestId("email_check").props.value).toBe(email);
  });

  //   //pass state changes by inputtext
  it("check pass input", async () => {
    const push = jest.fn();
    const { getByTestId, getByPlaceholderText } = await render(
      <Login navigation={{ push }} />
    );
    const pass = "salamThisis234";
    await fireEvent.changeText(getByTestId("pass_check"), pass);
    expect(getByTestId("pass_check").props.value).toBe(pass);
  });

  //   //empty email
  //     it("check empty email", async () => {
  //       const push = jest.fn();
  //       const { getByTestId, getByPlaceholderText } = await render(
  //         <Login navigation={{ push }} />
  //       );

  //       const empty_mail = "لطفا ایمیل خود را وارد کنید.";
  //       // showMessage.mockResolvedValue(empty_mail);

  //       FlashMessage.showMessage = jest.fn().mockResolvedValue(empty_mail);
  //       // jest.spyOn(FlashMessage, "showMessage").mockResolvedValueOnce(empty_mail);

  //       const email = "";
  //       console.log("this is email test length", email.length);
  //       await fireEvent.changeText(getByTestId("email_check"), email);
  //       await fireEvent.press(getByTestId("login-button"));
  //   expect(FlashMessage.showMessage).toHaveBeenCalled();
  //     // expect(getByTestId("pass_check").props.value).toBe(email);
  //     });

  // //empty password
  // it("check empty email", async () => {
  //   const push = jest.fn();
  //   const { getByTestId, getByPlaceholderText } = await render(
  //     <Login navigation={{ push }} />
  //   );
  //   const pass = "";
  //   await fireEvent.changeText(getByTestId("pass_check"), pass);
  //   expect(getByTestId("pass_check").props.value).toBe(pass);
  // });

  //empty email and password

  //authentication failed

  //auth successfull
});
