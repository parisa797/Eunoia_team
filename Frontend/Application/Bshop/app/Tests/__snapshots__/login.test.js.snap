// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Login component be tested renders correctly 1`] = `
<View
  accessible={true}
  focusable={true}
  onClick={[Function]}
  onResponderGrant={[Function]}
  onResponderMove={[Function]}
  onResponderRelease={[Function]}
  onResponderTerminate={[Function]}
  onResponderTerminationRequest={[Function]}
  onStartShouldSetResponder={[Function]}
  style={
    Object {
      "alignItems": "center",
      "backgroundColor": "#fff",
      "flex": 1,
      "justifyContent": "center",
      "marginTop": "-5%",
    }
  }
>
  <Image
    source={1}
    style={
      Object {
        "height": "50%",
        "marginBottom": "10%",
        "marginTop": "-5%",
        "width": "85%",
      }
    }
  />
  <View
    style={
      Object {
        "alignContent": "center",
        "alignItems": "center",
        "backgroundColor": "#f1f1f2",
        "borderRadius": 20,
        "height": 45,
        "marginBottom": 20,
        "width": "80%",
      }
    }
  >
    <TextInput
      allowFontScaling={true}
      onChangeText={[Function]}
      placeholder="ایمیل"
      placeholderTextColor="#000"
      rejectResponderTermination={true}
      style={
        Object {
          "color": "black",
          "flex": 1,
          "fontSize": 20,
          "height": 50,
          "marginLeft": 20,
          "padding": 10,
        }
      }
      testID="email_check"
      underlineColorAndroid="transparent"
      value=""
    />
  </View>
  <View
    style={
      Object {
        "alignContent": "center",
        "alignItems": "center",
        "backgroundColor": "#f1f1f2",
        "borderRadius": 20,
        "height": 45,
        "marginBottom": 20,
        "width": "80%",
      }
    }
  >
    <TextInput
      allowFontScaling={true}
      onChangeText={[Function]}
      placeholder="رمز عبور"
      placeholderTextColor="#000"
      rejectResponderTermination={true}
      secureTextEntry={true}
      style={
        Object {
          "color": "black",
          "flex": 1,
          "fontSize": 20,
          "height": 50,
          "marginLeft": 20,
          "padding": 10,
        }
      }
      testID="pass_check"
      underlineColorAndroid="transparent"
      value=""
    />
  </View>
  <View
    accessible={true}
    focusable={true}
    onClick={[Function]}
    onResponderGrant={[Function]}
    onResponderMove={[Function]}
    onResponderRelease={[Function]}
    onResponderTerminate={[Function]}
    onResponderTerminationRequest={[Function]}
    onStartShouldSetResponder={[Function]}
    style={
      Object {
        "alignItems": "center",
        "backgroundColor": "#b31414",
        "borderRadius": 20,
        "height": 50,
        "justifyContent": "center",
        "marginBottom": 20,
        "marginTop": 10,
        "opacity": 1,
        "width": "40%",
      }
    }
  >
    <Text
      style={
        Object {
          "color": "#fff",
          "fontSize": 20,
        }
      }
    >
      ورود
    </Text>
  </View>
  <Text
    onPress={[Function]}
    style={
      Object {
        "color": "#b31414",
        "fontSize": 20,
        "marginTop": -9,
      }
    }
  >
    <Text
      style={
        Object {
          "color": "black",
          "fontSize": 20,
        }
      }
    >
      حساب کاربری ندارید؟
    </Text>
    ثبت نام
  </Text>
</View>
`;
